Adaptive Open Loop Partitioning

PhoenixWorthVCD opened the pull request:
## Description

Adaptive open loop partitioning consists of assigning a specific refinement mode for each block instead of using the same refinement mode for all the blocks within the picture. The refinement mode is selected based on the cost difference between the predicted block depth and the above and below depths. Assuming a maximum of 3 above refinement depths (Pred-3) and 3 below refinement depths (Pred+3) can be considered to provide the same quality as all depths, the algorithm can select one of the 16 [0..15] refinement combinations illustrated in the following table.


Combination | Start depth (s_depth) | End depth (e_depth) | Refinement mode
-- | -- | -- | --
0 | -3 | 3 | Pred -3/+3
1 | -2 | 3 | Pred -2/+3
2 | -1 | 3 | Pred -1/+3
3 | 0 | 3 | Pred +3
4 | -3 | 2 | Pred -3/+2
5 | -2 | 2 | Pred -2/+2
6 | -1 | 2 | Pred -1/+2
7 | 0 | 2 | Pred +2
8 | -3 | 1 | Pred -3/+1
9 | -2 | 1 | Pred -2/+1
10 | -1 | 1 | Pred -1/+1
11 | 0 | 1 | Pred +1
12 | -3 | 0 | Pred -3
13 | -2 | 0 | Pred -2
14 | -1 | 0 | Pred -1
15 | 0 | 0 | Pred

Closes #789 

## Type of Change

New feature

## Author

@NaderMahdi

## Tests and performance

- BD rate difference (compared to master commit d647dc) of -0.02% relative to AOM, averaged over all resolutions
- Speed gain of 13.1% averaged over all resolutions
- Slope of -0.0012 
- No bitstream differences for M{1-8}
- No memory deviations
- No effect on 10-bit encoding

joelsole commented:
Thanks for consolidating the macros and squash the commits. 
Could you test 10-bit sequences with enc-mode 0 to be sure it remains ok?

PhoenixWorthVCD commented:
> Thanks for consolidating the macros and squash the commits.
> Could you test 10-bit sequences with enc-mode 0 to be sure it remains ok?

Tested 360p, 720p, and 1080p clips in M0: the proposed changes do not affect the 10-bit bitstreams.

grill05 commented:
@PhoenixWorthVCD 
> Speed gain of 13.1% averaged over all resolutions
Would this speed gain be more significant for higher resolutions (720/1080p) or lower ones (480p) ?

PhoenixWorthVCD commented:
> @PhoenixWorthVCD
> 
> > Speed gain of 13.1% averaged over all resolutions
> 
> Would this speed gain be more significant for higher resolutions (720p/1080p) or lower ones (360p/480p) ?

The speed gain is more significant for higher resolutions than lower resolutions.  The gain is most significant for 1080p clips.

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -1571,6 +1588,27 @@ EB_EXTERN EbErrorType nsq_prediction_shape(
+                }
+            }
+        }
+    }
```
@ttrigui 
seems that we might need a more efficient sorting algorithm now that we have several arrays to be sorted

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -843,6 +843,23 @@ EbErrorType signal_derivation_multi_processes_oq(
+        // the motion information information for NSQ is generated.
+        if (picture_control_set_ptr->pic_depth_mode <= PIC_ALL_C_DEPTH_MODE) {
+            if (MR_MODE || sc_content_detected || sequence_control_set_ptr->static_config.enable_hbd_mode_decision)
+                picture_control_set_ptr->enable_adaptive_ol_partitioning = 0;
```
need a config flag for this feature

PhoenixWorthVCD replied:
config flag added: "-olpd-refinement".  The feature can only be used for M0 right now.

hassount commented:
@PhoenixWorthVCD thank you for addressing the comments, can you please rebase squash all commits  to one and perform some sanity checks before merging this ?

PhoenixWorthVCD commented:
> @PhoenixWorthVCD thank you for addressing the comments, can you please rebase squash all commits to one and perform some sanity checks before merging this ?

I have rebased and squashed.  I checked the BD rate of certain clips and found similar BD rates to what I found during testing.

joelsole commented:
@PhoenixWorthVCD is this going to be extended to 10-bit (enable_hbd_mode_decision on)?

PhoenixWorthVCD commented:
> @PhoenixWorthVCD is this going to be extended to 10-bit (enable_hbd_mode_decision on)?

Yes, this will be extended to 10-bit.  The refinement will be moved from the MDC process to the EncDec process, and a new PR is expected this week.
